Source/JavaScriptCore/dfg/DFGLoopUnrollingPhase.cpp

```diff
@@ -0,0 +1,823 @@
+        for (uint32_t i = header->predecessors.size(); i--;) {
+            BasicBlock* predecessor = header->predecessors[i];
+            if (m_graph.m_cpsDominators->dominates(header, predecessor))
+                continue;
+
+            preHeader = predecessor;
+            ++preHeaderCount;
+        }
+
+        if (preHeaderCount != 1)
+            return false;
+
+        data.preHeader = preHeader;
+        return true;
+    }
+
+    bool locateTail(LoopData& data)
+    {
+        BasicBlock* header = data.header();
+        BasicBlock* tail = nullptr;
+
+        for (BasicBlock* predecessor : header->predecessors) {
+            if (!m_graph.m_cpsDominators->dominates(header, predecessor))
+                continue;
+
+            if (tail) {
+                dataLogLnIf(Options::verboseLoopUnrolling(), "Skipping loop with header ", *header, " since it contains two tails: ", *predecessor, " and ", *tail);
+                return false;
+            }
+
+            tail = predecessor;
+        }
+
+        if (!tail) {
+            dataLogLnIf(Options::verboseLoopUnrolling(), "Skipping loop with header ", *header, " since it has no tail");
+            return false;
+        }
+
+        // PreHeader                          PreHeader
+        //  |                                  |
+        // Header <---                        Header_0
+        //  |        |       unrolled to       |
+        //  |       Tail  =================>  Branch_0
+        //  |        |                         |
+        // Branch ----                        Tail_0
+        //  |                                  |
+        // Next                               ...
+        //                                     |
+        //                                    Header_n
+        //                                     |
+        //                                    Branch_n
+        //                                     |
+        //                                    Next
+        //
+        // FIXME: This is not supported yet. We should do it only if it's profitable.
+        if (!tail->terminal()->isBranch()) {
+            dataLogLnIf(Options::verboseLoopUnrolling(), "Skipping loop with header ", *header, " since it has a non-branch tail");
+            return false;
+        }
+
+        for (BasicBlock* successor : tail->successors()) {
+            if (data.loop->contains(successor))
+                continue;
+            data.next = successor;
+        }
+        data.tail = tail;
+        return true;
+    }
+
+    bool isSupportedConditionOp(NodeType op);
+    bool isSupportedUpdateOp(NodeType op);
+
+    ComparisonFunction comparisonFunction(Node* condition);
+    UpdateFunction updateFunction(Node* update);
+
+    bool identifyInductionVariable(LoopData& data)
+    {
+        Node* condition = data.condition();
+        auto isConditionValid = [&]() ALWAYS_INLINE_LAMBDA {
+            if (!isSupportedConditionOp(condition->op()))
```

Comment:
Should we check whether the condition is typed with `Int32Use`?

Reply:
```
for (i = initialValue; condition(i, operand); i = update(i, updateValue)) { ... }
```
Sorry, I know it's confusing. Actually, `condition` is the terminal of the tail. I think we want to ensure the condition left (`i`) and right (`operand`) are in `Int32Use`.

Since we also want to avoid the unpredictable (actual hard to predict) constant iterations with
```
for (let i = 0; i < 4; ) {
    ...
    if (...) {
         i += 1;
    }
}
```
we want to ensure the pattern that condition left (`i` here) is exactly the update `update(i, updateValue)`.

Reply:
In this case, `D@35` is the `condition` node, `D@31` is the `update` node, and `D@34` is the `operand` node.
```
 10  1 19:   D@31:< 1:->	ArithAdd(Int32:D@28, Int32:D@29, Int32|PureNum|NeedsNaNOrInfinity|UseAsOther|ReallyWantsInt, Int32, CheckOverflow, Exits, bc#16, ExitValid)
 11  1 19:   D@32:<!0:->	MovHint(Check:Untyped:D@31, MustGen, loc5, W:SideState, ClobbersExit, bc#16, ExitInvalid)
 12  1 19:   D@33:< 1:->	SetLocal(Int32:D@31, loc5(I<Int32>/FlushedInt32), W:Stack(loc5), bc#16, exit: bc#19, ExitValid)  predicting Int32
 13  1 19:   D@34:< 1:->	JSConstant(JS|PureNum|NeedsNaNOrInfinity|UseAsOther, NonBoolInt32, Int32: 4, bc#19, ExitValid)
 14  1 19:   D@35:< 1:->	CompareLess(Int32:D@31, Int32:D@34, Boolean|PureNum|NeedsNegZero|NeedsNaNOrInfinity|UseAsOther, Bool, bc#19, ExitValid)
 15  1 19:   D@36:<!0:->	Branch(KnownBoolean:D@35, MustGen, T:#1/w:10.000000, F:#2/w:1.000000, W:SideState, bc#19, ExitInvalid)
```